compiler/rustc_trait_selection/src/traits/error_reporting/type_err_ctxt_ext.rs

```diff
@@ -464,7 +464,26 @@ impl<'tcx> TypeErrCtxt<'_, 'tcx> {
                                 root_obligation,
                             )
                         } else {
-                            (trait_predicate, &obligation)
+                            let mut temp_trait_predicate = trait_predicate;
+                            let mut loop_code = obligation.cause.code().clone();
+                            loop{
+                                let (code, base) = loop_code.peel_derives_with_predicate();
+                                if matches!(loop_code, ObligationCauseCode::ImplDerivedObligation(ref c) if tcx.has_attr(c.impl_or_alias_def_id, sym::do_not_recommend)) {
+                                    if let Some(base) = base {
+                                        let code = code.clone();
+                                        obligation.cause.map_code(|_| code);
+                                        obligation.predicate = ty::PredicateKind::Clause(ty::ClauseKind::Trait(base.skip_binder().clone())).to_predicate(tcx);
+                                        temp_trait_predicate = base.clone();
+                                        trait_predicate = base;
+                                    }
+                                }
+                                if loop_code == *code {
+                                    break;
+                                } else {
+                                    loop_code = code.clone();
+                                }
+                            }
```

Comment:
Can we add a comment in front of this? When skimming this code it will be really hard to figure out what's going on without paying a lot of attention.

Reply:
It really should be extracted into a helper function, I think.

Reply:
I've fixed the formatting and extracted it into a helper function in https://github.com/rust-lang/rust/compare/eca373a073222883772da9c30616b3c746e3d6f2..f0279abd052e1ec4fa39ab71edbd7484dbc84c7b